### ISSUE TITLE
How to include functions in parameter links

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#1027]
I have sometimes wanted to say something like

```
foo.x = 4000 + 10 * np.sin(angle.c0)
```

where `angle.c0` is a parameter, but it fails (at creation time) with an error like:

```
sherpa In [7]: src.xpos = 4000 + 10 * np.cos(const1d.angle)
TypeError: loop of ufunc does not support argument 0 of type Parameter which has no callable cos method
```

or, for non-NumPy ufuncs:

```
sherpa In [24]: src.xpos = 4000 + 10 * math.cos(const1d.angle)
TypeError: must be real number, not Const1D
```

We need to lift the function into the parameter language. Here is a solution which needs testing, and then deciding how to handle multiple arguments (possible to make this support multiple arguments): 

```
class FunctionParameter(sherpa.models.parameter.CompositeParameter):
    """Store a function of a single argument.

    We need to only evalate the function when required.
    """

    @staticmethod
    def wrapobj(obj):
        if isinstance(obj, sherpa.models.parameter.Parameter):
            return obj
        return sherpa.models.parameter.ConstantParameter(obj)

    def __init__(self, arg, func):

        if not callable(func):
            raise ValueError("func argument is not callable!")

        self.arg = self.wrapobj(arg)
        self.func = func

        # Would like to be able to add the correct import symbol here
        lbl = _mklabel(func)
        lbl += f'{func.__name__}({self.arg.fullname})'
        sherpa.models.parameter.CompositeParameter.__init__(self, lbl,
                                                            (self.arg,))

    def eval(self):
        return self.func(self.arg.val)


def _mklabel(func):
    """Create a 'nice' symbol for the function.

    Is this worth it over just using __name__?

    Parameters
    ----------
    func : function reference

    Returns
    -------
    label : str
        A representation of the function.
    """

    if isinstance(func, np.ufunc):
        return 'np.'

    try:
        lbl = f'{func.__module__}.'
    except AttributeError:
        return ''

    if lbl == 'builtins.':
        return ''

    return lbl
```

With this, we can say (`xcm` is just the module where I placed the code):

```
sherpa In [16]: src.xpos = 4000 + 10 * xcm.FunctionParameter(angle.c0, np.cos)

sherpa In [17]: print(src)
gauss2d.src
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   src.fwhm     thawed           10  1.17549e-38  3.40282e+38
   src.xpos     linked       4005.4 expr: (4000 + (10 * np.cos(angle.c0)))
   src.ypos     thawed            0 -3.40282e+38  3.40282e+38
   src.ellip    frozen            0            0        0.999
   src.theta    frozen            0     -6.28319      6.28319    radians
   src.ampl     thawed            1 -3.40282e+38  3.40282e+38

sherpa In [18]: angle.c0 = np.pi / 4

sherpa In [19]: print(src)
gauss2d.src
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   src.fwhm     thawed           10  1.17549e-38  3.40282e+38
   src.xpos     linked      4007.07 expr: (4000 + (10 * np.cos(angle.c0)))
   src.ypos     thawed            0 -3.40282e+38  3.40282e+38
   src.ellip    frozen            0            0        0.999
   src.theta    frozen            0     -6.28319      6.28319    radians
   src.ampl     thawed            1 -3.40282e+38  3.40282e+38
```

We'd probably want a function which just wrapped the FunctionParameter call, which is not user friendly but simple. It would be great if we could say `4000 + 10 * np.cos(angle.c0)` but I can't think of a way to get this working.

# Check

For the `math.cos` case, to show it behaves similarly to `np.cos`:

```
sherpa In [25]: src.xpos = 4000 + 10 * xcm.FunctionParameter(angle.c0, math.cos)

sherpa In [26]: src.xpos.val
Out[26]: 4007.0710678118653

sherpa In [27]: print(src.xpos)
val         = 4007.0710678118653
min         = -3.4028234663852886e+38
max         = 3.4028234663852886e+38
units       =
frozen      = True
link        = (4000 + (10 * math.cos(angle.c0)))
default_val = 4007.0710678118653
default_min = -3.4028234663852886e+38
default_max = 3.4028234663852886e+38

sherpa In [28]: angle.c0 = 2

sherpa In [29]: print(src.xpos)
val         = 3995.8385316345284
min         = -3.4028234663852886e+38
max         = 3.4028234663852886e+38
units       =
frozen      = True
link        = (4000 + (10 * math.cos(angle.c0)))
default_val = 3995.8385316345284
default_min = -3.4028234663852886e+38
default_max = 3.4028234663852886e+38
```

**Comment by hamogu** [comment]
I like where this is going. 

I've had a problem like this in the past and I handled it by hiding all the computation in a custom model (so, to sherpa, it looked as if my model has a parameter that was tied 1:1 to angle, but internally, my `calc` method used `4000 + np.cos(angle)`. 

With that in mind, I wonder if it's worth the effort though. `model1.par1 = 4000 + 10 * np.cos(angle.c0)` would be really good, but as @DougBurke points out is not easy to achieve (*). The code above is kind of in the middle and as such I wonder if it's worth the effort. Non-Python users might find it too complicated and Python experts too limited. An alternative way to address this would be though documentation with an example along the lines of what I just described above. Just 😈 - advocate to prevent feature creep...

(*) Of course, sympy does exactly that. Makes me wonder if we should make a bridge to accept sympy function objects as model expressions... GSOC 2021?
